calico_node/startup.go

```diff
@@ -51,8 +59,239 @@ func main() {
 	// Make sure we have a global cluster ID set.
 	log.Info("Ensuring a cluster guid is set")
 	ensureClusterGuid(c)
+
+	hostname := os.GetEnv("NODENAME")
+	if hostname == nil {
+		hostname = os.GetEnv("HOSTNAME")
+	}
+	if hostname == nil {
+		hostname = os.Hostname()
+	}
+
+	if d := os.Getenv("DATASTORE_TYPE"); d == "kubernetes" {
+		f, err := os.OpenFile("startup.env", os.O_APPEND|os.O_WRONLY, 0600)
+		if err != nil {
+			panic(fmt.Sprintf("Error loading config: %s", err))
+		}
+		f.WriteString("export DATASTORE_TYPE=kubernetes\n")
+		f.WriteString("export HOSTNAME=%s\n" % hostname)
+	}
+
+	meta := api.NodeMetaData{Name: hostname}
+
+	fmt.Printf("Waiting for etcd connection...")
```

Comment:
We also need to move this WAIT_FOR_DATASTORE loop up above any code that actually uses the datastore.

There are a few bits above that depend on a datastore connection. e.g `ensureClusterGuid`

Would also make the code a bit cleaner to encapsulate this into a function call. e.g `waitForDatastore()`